CRM/Mailing/Page/View.php

```diff
@@ -165,4 +162,31 @@ public function run($id = NULL, $contactID = NULL, $print = TRUE, $allowID = FAL
     }
   }
 
+  public function getMailingID($id): void {
+    if (!empty($id) && !is_array($id)) {
+      $check = CRM_Core_DAO::singleValueQuery("SELECT id FROM civicrm_mailing WHERE id = %1", [
+        1 => [$id, 'String'],
```

Comment:
Fun fact. This seems to get false positive matches. I think it's because mysql (or maybe pear db) knows id is an int, so does some casting and so comparing id='02fe73ac3' will match on id 2.

Reply:
<sarcasm>Wonderful</sarcasm> thanks Dave

Reply:
🤪

Reply:
@demeritcowboy what if we did WHERE `CAST(id as CHAR)`  do you think that would be ok? would there be much of a performance hit?

Reply:
Unscientifically seems about 20% slower, but for just one viewing I don't think matters much. Anecdotally the page load time for emails is spent loading the 40 MB of images they've put in the email. And for a bot guessing urls it could bog you down other ways anyway.

So it works. Might want to code-comment it.

Longer-term beyond this PR, why do we have hashes here if you can just use the id? It's not preventing guessing if you can sequentially try ids. And these don't seem to be the full hash anyway so increase the chance of collisions.

Reply:
@demeritcowboy I think its because there is a setting that allows you have the id in the view url changes the id out for the mailing hash for reasons (i can't recall when that was added)

Reply:
That confuses me because if there's a setting that determines whether we are generating/expecting hashes vs ids, then falling back to hash even if the setting is off doesn't make sense right? Why can't we just use that setting to determine what we're expecting?

Reply:
I suppose what happens if you have sent some mailings then turn the setting on now you have emails in wild with both formats right?

Reply:
Yes but it seems like checking both defeats the purpose of the hash. It was intended to be one or the other, otherwise hash adds no value. https://issues.civicrm.org/jira/browse/CRM-14436

Reply:
My understanding is that this is the scenario the code is trying to accommodate. The option to hash mailing IDs is a checkbox or radio boolean in the admin UI.

Reply:
Well, for the purposes of this PR, using CAST will make it work, so can punt fixing the inconsistency to some other day.

Reply:
ok I have done the cast as and the other update requested

Reply:
Except now I see that it's this PR that introduces that inconsistency. I thought it was there before but it did only let you use one or the other depending on the setting. You get access denied if you try to use the other.

Reply:
So I think this idea can still work to solve the problem of existing numerical hashes out in the wild, but it would need to take the hash setting into account in getMailingID.

Reply:
@demeritcowboy I'm not sure how it is adding the inconsistency considering L94 and L104 are still within the if of the we have enabled hashes but maybe i am missing something your seeing

Reply:
Hmm I must have got something mixed up in my testing. It all runs as expected.
jenkins retest this please